xayn-ai/src/coi/point.rs

```diff
@@ -180,24 +225,52 @@ impl CoiPoint for PositiveCoi_v0_1_0 {
 
 #[cfg(test)]
 impl CoiPoint for PositiveCoi_v0_2_0 {
-    fn new(id: CoiId, point: Embedding, _viewed: Option<Duration>) -> Self {
+    fn new(
+        id: CoiId,
+        point: Embedding,
+        _key_phrases: Vec<KeyPhrase>,
+        _viewed: Option<Duration>,
+    ) -> Self {
         Self { id, point }
     }
 
     coi_point_default_impls! {}
 }
 
 impl CoiPoint for PositiveCoi {
-    fn new(id: CoiId, point: Embedding, viewed: Option<Duration>) -> Self {
+    fn new(
+        id: CoiId,
+        point: Embedding,
+        key_phrases: Vec<KeyPhrase>,
+        viewed: Option<Duration>,
+    ) -> Self {
+        debug_assert!(key_phrases.iter().all(KeyPhrase::is_valid));
+        debug_assert!(KeyPhrase::is_unique(&key_phrases));
         Self {
             id,
             point,
+            key_phrases,
             stats: CoiStats::new(viewed),
         }
     }
 
     coi_point_default_impls! {}
+}
+
+impl CoiPointKeyPhrases for PositiveCoi {
+    fn key_phrases(&self) -> &[KeyPhrase] {
+        self.key_phrases.as_slice()
+    }
 
+    fn swap_key_phrases(&mut self, mut candidates: Vec<KeyPhrase>) -> Vec<KeyPhrase> {
```

Comment:
Why we need this? and why we are returning the candidates, we are loosing `key_phrases`

Reply:
we may change the function signature later on, for now it is implemented this way to be able to iterate over the owned key phrases in the follow up pr.

not sure what you mean by losing the `key_phrases`? there is a memory swap of `key_phrases` and `candidates`, meaning that the previous `key_phrases` are actually returned and the `candidates` become the new key phrases.